panther_hardware_interfaces/src/gpio_driver.cpp

```diff
@@ -5,56 +5,282 @@ namespace panther_hardware_interfaces
+
+  builder.add_line_settings(offset, settings);
+  pin_info.offset = offset;
+}
+
+gpiod::line_settings GPIOController::generate_line_settings(const GPIOinfo & pin_info)
+{
+  auto settings = gpiod::line_settings();
+  settings.set_output_value(pin_info.init_value);
+  settings.set_direction(pin_info.direction);
+  settings.set_bias(pin_info.bias);
+
+  if (pin_info.direction == gpiod::line::direction::INPUT) {
+    settings.set_edge_detection(gpiod::line::edge::BOTH);
+    settings.set_debounce_period(std::chrono::milliseconds(gpio_debounce_period_));
+  }
+
+  return settings;
+}
+
+void GPIOController::change_control_pin_direction(GPIOpins pin, gpiod::line::direction direction)
+{
+  auto & pin_info = gpio_info_[pin];
+
+  if (pin_info.direction == direction) {
+    return;
+  }
+
+  pin_info.direction = direction;
+  gpiod::line_settings settings = generate_line_settings(pin_info);
+
+  auto line_config = gpiod::line_config();
+  line_config.add_line_settings(pin_info.offset, settings);
+
+  controll_request_->reconfigure_lines(line_config);
+}
+
+bool GPIOController::get_pin_value(GPIOpins pin)
+{
+  return gpio_info_[pin].value == gpiod::line::value::ACTIVE;
+}
+
+bool GPIOController::set_pin_value(GPIOpins pin, bool value)
+{
+  auto & pin_info = gpio_info_[pin];
+
+  if (pin == GPIOpins::WATCHDOG || pin_info.direction == gpiod::line::direction::INPUT) {
+    return false;
+  }
+
+  gpiod::line::value gpio_value = value ? gpiod::line::value::ACTIVE : gpiod::line::value::INACTIVE;
+  controll_request_->set_value(pin_info.offset, gpio_value);
+
+  std::this_thread::sleep_for(std::chrono::milliseconds(1));
+
+  if (controll_request_->get_value(pin_info.offset) != gpio_value) {
+    return false;
+  }
+
+  pin_info.value = gpio_value;
+  publish_gpio_state_callback(pin_info);
+
+  return true;
+}
+
+void GPIOController::gpio_monitor_on()
+{
+  for (GPIOpins pin = GPIOpins::WATCHDOG + 1; pin < GPIOpins::UNKNOWN; pin = pin + 1) {
+    gpio_info_[pin].value = controll_request_->get_value(gpio_info_[pin].offset);
+  }
 
-  watchdog_thread_ = std::make_unique<std::thread>([this]() {
-    watchdog_->request(
-      {"enable_motor_pins_application", gpiod::line_request::DIRECTION_OUTPUT, 0}, 1);
-    while (true) {
-      watchdog_->set_value(!watchdog_->get_value());
-      std::this_thread::sleep_for(std::chrono::milliseconds(10));
+  gpio_monitor_thread_enabled_ = true;
+  gpio_monitor_thread_ = std::make_unique<std::thread>(&GPIOController::monitor_async_events, this);
+}
+
+void GPIOController::monitor_async_events()
+{
+  auto edge_event_buffer = gpiod::edge_event_buffer(edge_event_buffer_size_);
+
+  struct pollfd pollfd;
+  pollfd.fd = controll_request_->fd();
+  pollfd.events = POLLIN;
+
+  while (gpio_monitor_thread_enabled_) {
+    auto ret = poll(&pollfd, 1, -1);
+
+    if (ret == -1) {
+      std::cerr << "Error waiting for edge events" << std::endl;
+      return;
+    }
```

Comment:
here we can set some flag, so that it will be possible to check if error occurred